### PR TITLE
[Ng Man Sing] Individual Project

### PR DESCRIPTION
**Pull request by NgManSing**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,189 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Scanner;
```
Comment (lamzf1998) [inline src/main/java/Duke.java]: Its good that you import only the necessary libraries required.

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,189 @@
...
+        StringBuilder dueDay = new StringBuilder();
+        for (int i = 0; i < slashElementIndex; i++) {
+            taskName.append(detailFragments[i]);
+        }
+        for (int i = slashElementIndex + 1; i < detailFragments.length; i++) {
+            dueDay.append(detailFragments[i]);
+        }
+        return new String[]{taskName.toString(), dueDay.toString()};
+    }
+
```
Comment (lamzf1998) [inline src/main/java/Duke.java]: The codes does not have any deep nesting complications.

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,189 @@
...
+        System.out.println("Here is your task List:");
+        for (int i = 0; i < records.size(); i++) {
+            System.out.println((i + 1) + ". " + records.get(i));
+        }
+    }
+
+    private static void quitProgram() {
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+}
```
Comment (lamzf1998) [inline src/main/java/Duke.java]: Spacings are very consistent and the method names are simple yet understandable.

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,189 @@
...
+            if (detailFragments[slashElementIndex].substring(1).equals("at")) {
+                isTypeMatch = true;
+            }
+            break;
+        }
+        if (!isTypeMatch) {
+            System.out.println("Invalid input! (keywords not matching)");
+            return null;
+        }
+
```
Comment (lamzf1998) [inline src/main/java/Duke.java]: No redundant methods are introduced.

**Review by lamzf1998 — COMMENTED** [review]
Well done, didn't find a lot of coding standard violations. Keep up the good work!

File: src/main/java/Duke.java (hunk 1 above)
Comment (NgManSing) [reply src/main/java/Duke.java]: Thanks!

File: src/main/java/Duke.java (hunk 4 above)
Comment (NgManSing) [reply src/main/java/Duke.java]: Thanks!

File: src/main/java/Duke.java (hunk 2 above)
Comment (NgManSing) [reply src/main/java/Duke.java]: Thanks!

File: src/main/java/Duke.java (hunk 3 above)
Comment (NgManSing) [reply src/main/java/Duke.java]: Thanks!

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,189 @@
...
+            case "done":
+                executeCommandDone(arguments);
+                break;
+            case "bye":
+                isLoop = executeCommandBye(rawInput, arguments);
+                break;
+            default:
+                promptUserInputInvalid();
+                break;
+            }
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: Varying levels of abstraction.

Perhaps it will be great if you can shift line 19 to 22 and the switch statement to a separate method each, so as to achieve "Single Level of Abstraction Principle".

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -1,10 +1,189 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Scanner;
+
+/**
+ * This class is to build a personal assistance chat-bot called "Kaman"
+ * (Customised from Duke)
```
Comment (gerardtwk) [inline src/main/java/Duke.java:7]: The word "Customised" is a British spelling and violates the coding standard of using American spelling only. 
Perhaps you may want to correct it to "Customized" instead.

File: src/main/java/Duke.java [hunk 7]
```diff
@@ -1,10 +1,191 @@
...
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: The word "Customised" is a British spelling and violates the coding standard of using American spelling only.
Perhaps you may want to correct it to "Customized" instead.

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -1,10 +1,191 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Scanner;
+
+/**
+ * This class is to build a personal assistance chat-bot called "Kaman"
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: I like how you personalize your chatbot with a name. 

I think there is a minor wording issue with the phrase "personal assistance chat-bot" in your comment. Perhaps you might want to rephrase it to "personal assistant chatbot".

File: src/main/java/Duke.java (hunk 1 above)
Comment (gerardtwk) [reply src/main/java/Duke.java]: I agree on this too! The way you list your imports explicitly allows future collaborators to know exactly what libraries they need to be aware of before working on your code.

File: src/main/java/Deadline.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
+public class Deadline extends Task {
+
+    public static final String TASK_TYPE = "D";
```
Comment (gerardtwk) [inline src/main/java/Deadline.java]: I like the way you standardize the declaration of the constant TASK_TYPE across other subclasses of Task.

The way you name, capitalize and declare the constant TASK_TYPE adheres very well to the coding standard and convention, well done!

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,191 @@
...
+            break;
+        case Event.TASK_TYPE:
+            details = getTaskNameAndDate(detailFragments, taskType);
+            if (details != null) {
+                taskName = details[0];
+                date = details[1];
+                records.add(new Event(taskName, date));
+            }
+            break;
+        }
```
Comment (gerardtwk) [inline src/main/java/Duke.java:127]: I believe a "default" case should be added here, so as to adhere to the coding standard on using switch statement.

File: src/main/java/Duke.java [hunk 11]
```diff
@@ -1,10 +1,191 @@
...
+            }
+            break;
+        case Event.TASK_TYPE:
+            details = getTaskNameAndDate(detailFragments, taskType);
+            if (details != null) {
+                taskName = details[0];
+                date = details[1];
+                records.add(new Event(taskName, date));
+            }
+            break;
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: I believe a "default" case should be added here, so as to adhere to the coding standard on using a switch statement.

File: src/main/java/Duke.java [hunk 12]
```diff
@@ -1,10 +1,191 @@
...
+                break;
+            }
+        } while (isLoop);
+    }
+
+    private static void promptUserInputInvalid() {
+        System.out.println("Invalid command! Please try again!");
+    }
+
+    private static boolean executeCommandBye(String rawInput, String[] arguments) {
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: The naming of this method can be further improved as the current name does not sound like a boolean. This violates the coding standard of "Boolean variables/methods should be named to sound like booleans".

File: src/main/java/Duke.java [hunk 13]
```diff
@@ -1,10 +1,191 @@
...
+        case Deadline.TASK_TYPE:
+            if (detailFragments[slashElementIndex].substring(1).equals("by")) {
+                isTypeMatch = true;
+            }
+            break;
+        case Event.TASK_TYPE:
+            if (detailFragments[slashElementIndex].substring(1).equals("at")) {
+                isTypeMatch = true;
+            }
+            break;
```
Comment (gerardtwk) [inline src/main/java/Duke.java]: I believe a "default" case should be added here, so as to adhere to the coding standard on using a switch statement.

**Review by gerardtwk — COMMENTED** [review]
Overall code is readability is alright, just need to take note of the minor issues pointed out in my comments. Good job and keep up the good work!👍

File: src/main/java/duke/Duke.java [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+        case "list":
+            showList(userInput.getOtherArguments());
+            break;
+        case "done":
+            processCommand(userInput.getOtherArguments(), CommandType.done);
+            break;
+        case "delete":
+            processCommand(userInput.getOtherArguments(), CommandType.delete);
+            break;
+        case "bye":
```
Comment (wamikamalik) [inline src/main/java/duke/Duke.java]: Can consider replacing these magic strings(like bye, delete, etc) with constants.

File: src/main/java/duke/Duke.java [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+        switch (taskType) {
+        case Todo.TASK_TYPE:
+            if (detailFragments.length > 0) {
+                taskName = String.join(" ", detailFragments);
+                records.add(new Todo(taskName));
+                isAdded = true;
+            } else {
+                showInvalidEmptyDescription();
+            }
+            break;
```
Comment (wamikamalik) [inline src/main/java/duke/Duke.java]: Your method is a little long. You can consider refactoring some parts of your case clauses. See [avoid long methods](https://nus-cs2113-ay2021s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods)

File: src/main/java/duke/Duke.java [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+        }
+        return isAdded;
+    }
+
+    private static void showInvalidEmptyDescription() {
+        System.out.println("The description of a task cannot be empty.");
+    }
+
+    private static String[] getTaskNameAndDate(String[] detailFragments, String taskType) {
+        int keywordIndex = Arrays.asList(detailFragments).indexOf(taskType.equals(Deadline.TASK_TYPE) ? "/by" : "/at");
```
Comment (wamikamalik) [inline src/main/java/duke/Duke.java]: You can consider splitting this statement to improve readability.

File: src/main/java/duke/Input/CommandType.java [hunk 17]
```diff
@@ -0,0 +1,5 @@
+package duke.Input;
+
+public enum CommandType {
+    delete, done;
+}
```
Comment (wamikamalik) [inline src/main/java/duke/Input/CommandType.java]: Great use of enum! 👍 But remember to rename the constants in uppercase. For example
`DELETE, DONE;` Similar issue found in the other enum.

File: src/main/java/duke/Duke.java [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+
+    private static boolean isEndProgram(String[] arguments) {
+        if (arguments.length != 0) {
+            System.out.println("Command \"bye\" requires no argument. Please try again!");
+            return true;
+        }
+
+        quitProgram();
+        return false;
+    }
```
Comment (wamikamalik) [inline src/main/java/duke/Duke.java]: I like that you have refactored your code to make it more readable.

File: src/main/java/duke/Input/CommandType.java [hunk 19]
```diff
@@ -0,0 +1,5 @@
+package duke.Input;
+
+public enum CommandType {
```
Comment (wamikamalik) [inline src/main/java/duke/Input/CommandType.java]: Consider adding the other command types as well.
